### PR TITLE
feat(systest): check unallowed log patterns on the local backend

### PR DESCRIPTION
**Pull request by basvandijk**
## Summary

The `assert_no_unallowed_log_patterns` teardown (`"panicked"`, `"This is a bug"`, ...) queried ElasticSearch and was gated on `logs_enabled`, so it never ran on the **local** backend: the `<name>_local` targets always pass `--no-logs` because the sandbox has no network and no Vector VM.

This PR makes the check work on the local backend by scanning the journald records that `logs_stream_task` already streams:

- **`logs_stream_task`** now also persists every streamed target's records as JSON Lines under `<group_dir>/journald_logs/{nodes,uvms}/<target>.jsonl` (a sibling of the env dirs, which are `cp -R`'d per test). Each record is one unbuffered `write_all` on an `O_APPEND` file, because the streamer is SIGKILLed with no flush window and the teardown scans while it is still appending. `JournalRecord` gained `__REALTIME_TIMESTAMP` and `Serialize`. Node discovery now also includes API boundary nodes.
- **New `unallowed_log_patterns` module** holding the whole check. A `LogSource` chosen in `make_plan` selects ElasticSearch on Farm (unchanged query) or the `nodes/` files on the local backend (requires `--stream-ic-node-logs`, which the `_local` variant always sets; only `nodes/` is scanned, mirroring the ES `ic` filter which only covers IC nodes). Both sources share:
  - one **Vector-style `MESSAGE` normalizer** (replica slog lines are matched on the inner `log_entry.message`, exactly what ES stores, so exclusions like `canister` cannot accidentally match JSON metadata such as crate names),
  - one **ASCII case-insensitive substring predicate** (ES `match_phrase` lowercases too; on Farm this turns the confusing "N hit(s), but none could be attributed" report for lowercase `this is a bug` emitters into a proper report, the hit set is unchanged),
  - one report/panic renderer (same text and escape hatches as before).
- The local scan skips a truncated final line, counts unparseable lines, de-duplicates matches by `__CURSOR` (a reconnecting stream re-emits the record at its cursor), waits 5 s so records from the last moments of the test land on disk, and fails open like the ES path (missing dir / no records => logged and skipped, with a hint whether an IC was deployed at all).
- Docs: CLI help for `--no-logs` / `--exclude-logs` / `--stream-ic-node-logs`, builder docs, group-dir layout in `constants.rs`, stale `SYSTEM_TEST_INFRA` in `local_backend.rs`.

## Verification

- `cargo check`, `cargo fmt`, `./ci/scripts/rust-lint.sh` clean; `bazel test //rs/tests/driver:unit_tests //rs/tests/idx:test_e2e_scenarios` pass (28 new unit tests: record (de)serialization and round trip, sink, path layout, normalizer table, predicate incl. case-insensitivity, report cap, scan edge cases, end-to-end soft-skip and `#[should_panic]`).
- `//rs/tests/idx:basic_health_test_local` **passes**; the teardown logged `Scanned 13744 journald record(s) (4680649 byte(s)) in 8 file(s) ... 0 truncated line(s), 0 unparseable line(s), 0 duplicate match(es)` and the 8 `nodes/<node_id>.jsonl` files have no duplicate cursors.
- Positive path: temporarily adding `.add_unallowed_log_pattern("Reached target")` and `.add_unallowed_log_pattern("reached TARGET")` to `basic_health_test` makes only the teardown fail, both patterns report the same 160 systemd lines with `[<rfc3339> <node_id>] <message>` samples (patch reverted).

## Notes

- Every `_local` target now gets this teardown in the nightly `local-system-tests` workflow. Tests that deliberately crash or kill replicas (`replica_determinism_test_local`, `delete_subnet_test_local`, the malicious-replica consensus tests, `kill_start_*_test_local`, `global_reboot_test_local`) are the likeliest to surface newly caught panics; the existing `add_unallowed_log_pattern_except` escape hatch applies.
- Known limitations (documented in the module): IC nodes excluded via `--exclude-logs` are not scanned locally; nodes registering mid-test are not discovered (setup-time topology snapshot, pre-existing); substring matching is not token-anchored like `match_phrase`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rs/tests/driver/src/driver/unallowed_log_patterns.rs [hunk 1]
```diff
@@ -0,0 +1,1114 @@
...
+            log_entry
+                .get("message")
+                .and_then(serde_json::Value::as_str)
+                .unwrap_or_default()
+                .to_string(),
+        ),
+        _ => match object.get("message") {
+            Some(serde_json::Value::String(message)) => Cow::Owned(message.clone()),
+            _ => Cow::Borrowed(raw),
+        },
```
[the remark this answers is not in the document]
Comment (basvandijk) [reply rs/tests/driver/src/driver/unallowed_log_patterns.rs:376]: Fixed in 198d90afd0: a JSON log line without a string top-level `message` now yields Vector's `"Log contained no message field"` sentinel (missing/null) or an unmatchable empty string (non-string), and the raw JSON is never matched. Covered by `normalize_message_flat_json_without_string_message_is_never_the_raw_json`.

File: rs/tests/driver/src/driver/unallowed_log_patterns.rs [hunk 2]
```diff
@@ -0,0 +1,1114 @@
...
+/// Scans one persisted journald log file (see `logs_stream_task::JournalFileSink`)
+/// line by line, in bounded memory. The writer may still be appending: a final
+/// line without a trailing newline is its in-flight write and is skipped.
+fn scan_journald_log_file(
+    path: &Path,
+    node: &str,
+    patterns: &CompiledPatterns,
+    out: &mut ScanOutcome,
+) -> std::io::Result<()> {
+    let mut reader = BufReader::with_capacity(64 * 1024, std::fs::File::open(path)?);
```
[the remark this answers is not in the document]
Comment (basvandijk) [reply rs/tests/driver/src/driver/unallowed_log_patterns.rs]: Fixed in 198d90afd0: `scan_journald_log_file` snapshots the file length at open and reads through `Read::take(len)`, so a growing file cannot keep the scan alive; a line cut at the snapshot boundary is skipped as truncated. Covered by `scan_reader_treats_a_line_cut_at_the_byte_limit_as_truncated`.

File: rs/tests/driver/src/driver/logs_stream_task.rs [hunk 3]
```diff
@@ -42,6 +49,44 @@ pub(crate) const LOGS_STREAM_TASK_NAME: &str = "logs_stream";
 const RETRY_DELAY_JOURNALD_STREAM: Duration = Duration::from_secs(5);
 const RETRY_DELAY_DISCOVER_TARGETS: Duration = Duration::from_secs(5);
 
+/// `<group_dir>/journald_logs`: the persisted journald records, one JSON Lines
+/// file per streamed target (see [`JournalFileSink`]). A sibling of
+/// `root_env/`, `setup/` and `tests/` because those env directories are
+/// `cp -R`'d into every per-test directory (`TestEnv::fork_from`), and
+/// deliberately not under `local_backend/` because this task is
+/// backend-agnostic and runs on Farm too.
```
Comment (basvandijk) [inline rs/tests/driver/src/driver/logs_stream_task.rs]: To much detail, I would suggest:
```suggestion
/// `<group_dir>/journald_logs`: the persisted journald records, one JSON Lines
/// file per streamed target (see [`JournalFileSink`]).
```

**Comment by basvandijk** [comment]
`local-system-tests ` run: https://github.com/dfinity/ic/actions/runs/34327146029/job/102386921496.

**Review by dfinityianblenke — APPROVED** [review]
